File: 0645-set-mismatch/0645-set-mismatch.cpp
class Solution {
public:
    vector<int> findErrorNums(vector<int>& nums) {
        int n = nums.size();
        int xorAll = 0;
        int xorArray = 0;
        
        
        // 1 ~ n xor
        for (int i = 1; i <= n; i++) {
            xorAll ^= i;
        }
        
        // given nums xor
        for (int num : nums) {
            xorArray ^= num;
        }
        
        int xorResult = xorArray ^ xorAll;
        
        // 최상위 비트
        int rightmostSetBit = xorResult & -xorResult;
        
        int xorSet = 0;
        int xorNotSet = 0;
        
        // 추가 
        for (int i = 1; i <= n; i++) {
            if (i & rightmostSetBit) {
                xorSet ^= i;
            } else {
                xorNotSet ^= i;
            }
        }
        
        //제거
        for (int num : nums) {
            if (num & rightmostSetBit) {
                xorSet ^= num;
            } else {
                xorNotSet ^= num;
            }
        }
        
        for (int num : nums) {
            // 있다면 중복된 녀석이 xorSet
            if (num == xorSet) {
                return {xorSet, xorNotSet};
            }
        }
        // 없다면 xorNotSet 이 중복된 녀석
        return {xorNotSet, xorSet};
    }
};

File: 0645-set-mismatch/0645-set-mismatch.java
class Solution {
    public int[] findErrorNums(int[] nums) {

        int n = nums.length;
        int[] arr = new int[n+1];
        int[] res = new int[2];
        
        for (int i = 0; i < n; i++) {
            arr[nums[i]]++;            
        }
        
        for (int i = 1; i <= n; i++) {
            if (arr[i] == 2) {
                res[0] = i;
            } else if (arr[i] == 0) {
                res[1] = i;
            }
        }

        return res;
    }
}